### PR TITLE
Rdi 2212

### PR DESCRIPTION
**Pull request by turran**

### CODE REVIEW

File: gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportsrc.cpp [hunk 1]
```diff
@@ -487,6 +534,10 @@ gst_web_transport_src_handle_message (GstBin *bin, GstMessage *msg)
   GST_LOG_OBJECT (bin, "Handle message of type %s",
       gst_message_type_get_name (GST_MESSAGE_TYPE (msg)));
 
+  if ((handled = gst_web_transferable_handle_request_object (
+           (GstWebTransferable *) bin, msg)))
```
Comment (cfoch) [inline gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportsrc.cpp:538]: I would add it inside the switch statement just for readability inside a `case GST_MESSAGE_ELEMENT`.

**Review by cfoch — CHANGES_REQUESTED** [review]

File: gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportstreamsrc.cpp [hunk 2]
```diff
@@ -86,19 +95,20 @@ gst_web_transport_stream_src_create (GstPushSrc *psrc, GstBuffer **outbuf)
...
-    m = gst_web_utils_message_new_request_object (GST_ELEMENT_CAST (self),
-        "gst_web_transport_stream_src_on_stream", object_name, self);
-    g_free (object_name);
+    gst_web_transferable_register_on_message ((GstWebTransferable *) self);
     /* Request the WebTransportReceiveStream */
-    GST_DEBUG_OBJECT (self, "Requesting stream %s", self->name);
-    if (!gst_element_post_message (GST_ELEMENT_CAST (self), m)) {
-      GST_ERROR_OBJECT (self, "Posting message failed");
+    requested =
+        gst_web_transferable_request_object ((GstWebTransferable *) self,
```
[the remark this answers is not in the document]
Comment (turran) [reply gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportstreamsrc.cpp:107]: This was on purpose. From the PoV of the element, you just call one function and that's it. This narrows down the logic an element must implement which ends to be very repetitive among elements.

From the PoV of the implementation, it is fine to do different kind of processing, similar to how GstContexts are managed. You try by sending a query/event, if it is not answered, you send a message to the bus.

File: gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    return FALSE;
+
+  object_name = gst_structure_get_string (s, "object-name");
+  GST_DEBUG_OBJECT (self,
+      "Handling the request of object to '%s' to element %s'", object_name,
+      GST_OBJECT_NAME (GST_MESSAGE_SRC (msg)));
+
+  /* check if the rquested object is one of the transferable objects */
+  iface = GST_WEB_TRANSFERABLE_GET_INTERFACE (self);
+  if (iface->can_transfer && iface->can_transfer (self, object_name)) {
```
[the remark this answers is not in the document]
Comment (turran) [reply gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp:254]: A g_return_val_if_fail would make a GLib assertion if the pre-condition is not met. Not implementing such a method is not mandatory. I wouldn't assert if it is not implemeted.

File: gst.wasm/subprojects/samples/template.html [hunk 4]
```diff
@@ -27,20 +27,22 @@
     <script type="text/javascript">
       var Module = {};
 
-      function onPageLoad() {
+      function onPageLoad(o) {
```
[the remark this answers is not in the document]
Comment (turran) [reply gst.wasm/subprojects/samples/template.html:30]: Ok

File: gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    /* if so, transfer it */
+    if (iface->transfer) {
+      iface->transfer (self, object_name, msg);
```
[the remark this answers is not in the document]
Comment (turran) [reply gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp:257]: Reviewing the code, yes, I wanted it to return a boolean but there is no code that can actually fail on that case, at least for now. That's why there is a `can_transfer` call before in the interface code, not the element.

File: gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportsrc.cpp (hunk 1 above)
Comment (turran) [reply gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportsrc.cpp:538]: This is on purpose. Adding a switch statement on the element makes the element aware of the potential messages the interface can handle for you, which breaks the contract of having the logic split.  What if the interface handles a message that is not a `GST_MESSAGE_ELEMENT`, then you need to add another `case` which is pointless. Let the interface handle the message (of whatever kind) and check the return value, that's all.

File: gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp [hunk 6]
```diff
@@ -285,14 +285,16 @@ gst_web_transferable_transfer_object (
...
-      {
-        gst_web_transferable_js_worker_transfer_object (Emval.toValue ($0),
-            Emval.toValue ($1), Emval.toValue ($2), $3, $4, $5);
-      },
-      val::u8string (cb_name).as_handle (),
-      val::u8string (object_name).as_handle (), object, user_data, tid,
-      pthread_self ());
+  /* clang-format off */
+  EM_ASM ({
+    gst_web_transferable_js_worker_transfer_object (Emval.toValue ($0),
```
Comment (cfoch) [inline gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp:290]: This formatting is not too bad to turn off clang-format, IMO.

File: gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+
+#include <gst/gst.h>
+#include <emscripten.h>
+#include <emscripten/bind.h>
+#include <emscripten/val.h>
+
+#include "gstwebtransferable.h"
+
+#define GST_CAT_DEFAULT gst_web_transferable_debug
+#define GST_WEB_TRANSFERABLE_MESSAGE_REQUEST_OBJECT_NAME                      \
```
Comment (cfoch) [inline gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp:33]: Isn't this an opportunity to use a GQuark?

File: gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (cfoch) [reply gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp:254]: Ok, I agree.

File: gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+  g_return_val_if_fail (GST_IS_WEB_TRANSFERABLE (self), FALSE);
+
+  /* check if the message is of the type we expect */
+  if (GST_MESSAGE_TYPE (msg) != GST_MESSAGE_ELEMENT)
+    return FALSE;
+
+  s = gst_message_get_structure (msg);
+  if (!s)
+    return FALSE;
+  if (g_strcmp0 (gst_structure_get_name (s),
```
Comment (cfoch) [inline gst.wasm/subprojects/gst-plugins-web/gst-libs/gst/web/gstwebtransferable.cpp:243]: If we use a GQuark for GST_WEB_TRANSFERABLE_MESSAGE_REQUEST_OBJECT_NAME there is no need for using strcmp.

File: gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportsrc.cpp (hunk 1 above)
Comment (cfoch) [reply gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportsrc.cpp:538]: Ok, that makes sense.

File: gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportstreamsrc.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (cfoch) [reply gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportstreamsrc.cpp:107]: It is still hard to follow though. A comment on the function would be useful.

**Review by cfoch — APPROVED** [review]
Just requesting for the comment. But is so simple change, that I will approve in advance.

Consider the rest of comments as desires/nitpicks. Feel free to implement them or not.

File: gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportstreamsrc.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (turran) [reply gst.wasm/subprojects/gst-plugins-web/gst/web/transport/gstwebtransportstreamsrc.cpp:107]: Alright
